File: Problems/034-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    # returns leftmost (or rightmost) index at which `target` should be inserted in sorted
    # array `nums` via binary search.
    def extreme_insertion_index(self, nums, target, flag):
        left = 0
        right = len(nums)

        while left < right:
            mid = (left + right) // 2
            if nums[mid] > target or (flag and target == nums[mid]):
                right = mid
            else:
                left = mid + 1

        return left

    def searchRange(self, nums, target):
        left_idx = self.extreme_insertion_index(nums, target, True)

        # assert that `left_idx` is within the array bounds and that `target`
        # is actually in `nums`.
        if left_idx == len(nums) or nums[left_idx] != target:
            return [-1, -1]

        return [
            left_idx,
            self.extreme_insertion_index(nums, target, False) - 1
        ]

File: Problems/034-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.go
package problem034

func searchRange(nums []int, target int) []int {
	leftIdx := binarrySearch(nums, target, true)

	if leftIdx == len(nums) || nums[leftIdx] != target {
		return []int{-1, -1}
	}

	return []int{leftIdx, binarrySearch(nums, target, false) - 1}
}

func binarrySearch(nums []int, target int, flag bool) int {
	left, right := 0, len(nums)
	var mid int
	for left < right {
		mid = (left + right) >> 1
		if nums[mid] > target || (flag && nums[mid] == target) {
			right = mid
		} else {
			left = mid + 1
		}
	}

	return left
}
